File: src/main/java/me/testaccount666/migration/essentials/HomeMigrator.java
package me.testaccount666.migration.essentials;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.userdata.OfflineUser;

import java.util.logging.Level;


public class HomeMigrator extends AbstractMigrator {

    @Override
    public int migrateFrom() {
        var count = 0;

        var userManager = userManager();
        var essentials = essentials();
        var uuids = essentials.getUsers().getAllUserUUIDs();

        for (var uuid : uuids) {
            var userOptional = userManager.getUser(uuid);
            if (userOptional.isEmpty()) {
                ServerSystem.getLog().warning("Couldn't find user '${uuid}', skipping home migration!");
                continue;
            }

            count += migrateFrom(userOptional.get().getOfflineUser());
        }

        return count;
    }

    private int migrateFrom(OfflineUser user) {
        var count = 0;

        var homeManager = user.getHomeManager();
        var essentials = essentials();

        var essentialsUser = essentials.getUser(user.getUuid());
        for (var homeName : essentialsUser.getHomes())
            try {
                var location = essentialsUser.getHome(homeName);

                homeManager.addHome(homeName, location);
                count += 1;
            } catch (IllegalArgumentException exception) {
                var userName = user.getName().orElse("Unknown");

                ServerSystem.getLog().log(Level.WARNING, "Couldn't migrate home '${homeName}' for user '${user.getUuid()}' (${userName})", exception);
            }

        user.save();

        return count;
    }

    @Override
    public int migrateTo() {
        var count = 0;

        var userManager = userManager();

        for (var player : offlinePlayers()) {
            var uuid = player.getUniqueId();

            var userOptional = userManager.getUser(uuid);
            if (userOptional.isEmpty()) {
                ServerSystem.getLog().warning("Couldn't find user '${uuid}', skipping home migration!");
                continue;
            }

            count += migrateTo(userOptional.get().getOfflineUser());
        }

        return count;
    }

    private int migrateTo(OfflineUser user) {
        var count = 0;

        var homeManager = user.getHomeManager();
        var essentials = essentials();

        ensureUserDataExists(user.getUuid());
        var essentialsUser = essentials.getUser(user.getUuid());

        for (var home : homeManager.getHomes()) {
            var homeName = home.getDisplayName();
            var location = home.getLocation();

            try {
                essentialsUser.setHome(homeName, location);

                count += 1;
            } catch (Exception exception) {
                var userName = user.getName().orElse("Unknown");

                ServerSystem.getLog().log(Level.WARNING, "Couldn't migrate home '${homeName}' for user '${user.getUuid()}' (${userName})", exception);
            }
        }

        return count;
    }
}

File: src/main/kotlin/me/testaccount666/migration/plugins/essentials/HomeMigrator.kt
package me.testaccount666.migration.plugins.essentials

import me.testaccount666.serversystem.ServerSystem.Companion.log
import me.testaccount666.serversystem.userdata.OfflineUser
import java.util.logging.Level

class HomeMigrator : AbstractMigrator() {
    override fun migrateFrom(): Int {
        val count = essentials.users.allUserUUIDs.sumOf {
            val cachedUser = userManager.getUserOrNull(it) ?: run {
                log.warning("Couldn't find user '${it}', skipping home migration!")
                return@sumOf 0
            }

            return@sumOf migrateFrom(cachedUser.offlineUser)
        }

        return count
    }

    private fun migrateFrom(user: OfflineUser): Int {
        val essentialsUser = essentials.getUser(user.uuid)
        val count = essentialsUser.homes.count { homeName ->
            runCatching {
                val location = essentialsUser.getHome(homeName)

                user.homeManager.addPoint(homeName, location)
                return@runCatching true
            }.onFailure {
                log.log(Level.WARNING, "Couldn't migrate home '${homeName}' for user '${user.uuid}' (${user.nameSafe})", it)
            }.getOrDefault(false)
        }

        user.save()

        return count
    }

    override fun migrateTo(): Int {
        val count = offlinePlayers().sumOf {
            val uuid = it.uniqueId

            val cachedUser = userManager.getUserOrNull(uuid) ?: run {
                log.warning("Couldn't find user '${uuid}', skipping home migration!")
                return@sumOf 0
            }

            return@sumOf migrateTo(cachedUser.offlineUser)
        }

        return count
    }

    private fun migrateTo(user: OfflineUser): Int {
        ensureUserDataExists(user.uuid)
        val essentialsUser = essentials.getUser(user.uuid)

        val count = user.homeManager.waypoints.count { home ->
            runCatching {
                val homeName = home.displayName
                val location = home.location
                essentialsUser.setHome(homeName, location)
                return@runCatching true
            }.onFailure {
                log.log(Level.WARNING, "Couldn't migrate home '${home.displayName}' for user '${user.uuid}' (${user.nameSafe})", it)
            }.getOrDefault(false)
        }

        return count
    }
}
